Implement `RetAbi` and `BoxRet`

workingjubilee opened the pull request:
This completely reworks the way we handle returning types from functions, so that we no longer have to rely on a macro expansion behavior that has to already know the types at expansion time (and thus has to parse them somehow, which it cannot realistically do because type aliases exist). Instead, we simply expand into code that asks the types themselves to modify the FunctionCallInfo appropriately and then return a raw Datum to Postgres.

This breaks support for certain returns because it was difficult to do this and also support arbitrary nesting, because Postgres does not support arbitrary nesting. For instance, you can no longer return:
- `SetOfIterator<'a, Result<T, E>>`
- `TableIterator<'a, (Result<T, E>, Result<U, D>)>`
- `Option<TableIterator<'a, Tuple>>`

It's expected that this will improve in the near-ish future.

This also breaks returning values from `#[pg_extern]` functions that were relying on `IntoDatum` implementations being enough. It is not expected this will improve, for the reasons described on the documentation of the new traits, which can be summarized as "`IntoDatum` should never have been used for that bound". This change blocks off several latent correctness problems from affecting pgrx going forward.

Fixes https://github.com/pgcentralfoundation/pgrx/issues/1484

workingjubilee committed:
partial

workingjubilee committed:
another partial

workingjubilee committed:
Refactor a bunch of stuff

workingjubilee committed:
extract a wrapper for srfs

workingjubilee committed:
start work on the wrapper fn

workingjubilee committed:
check compiles proc macros but does not finish building pg_externs

workingjubilee committed:
check passes

workingjubilee committed:
almost builds all tests

workingjubilee committed:
explicitly impl BoxRet for &str

workingjubilee committed:
type inference defeated

workingjubilee committed:
Everything except composite types

workingjubilee committed:
Tests build and 8 tests fail

    tests::heap_tuple::tests::pg_test_tuple_desc_clone
    tests::result_tests::tests::pg_test_return_result_set_of
    tests::result_tests::tests::pg_test_return_result_set_of_error
    tests::srf_tests::tests::pg_test_generate_series
    tests::srf_tests::tests::pg_test_return_none_setof_iterator
    tests::srf_tests::tests::pg_test_return_some_setof_iterator
    tests::srf_tests::tests::pg_test_srf_setof_datum_detoasting_with_borrow
    tests::struct_type_tests::tests::pg_test_complex_storage_and_retrieval

workingjubilee committed:
silence deprecations

workingjubilee committed:
remove nonsense

workingjubilee committed:
ignore unused

workingjubilee committed:
fix 7 tests

workingjubilee committed:
fix last tests

workingjubilee committed:
Convert scalar values to single return strat

workingjubilee committed:
and unit AKA void returns

workingjubilee committed:
its wrapper_code now

workingjubilee committed:
unused

workingjubilee committed:
merge non-len-1 tableiterators in

workingjubilee committed:
document some safety preconditions before we forget

workingjubilee committed:
spread required unsafe impls around

workingjubilee committed:
layer tuple in

workingjubilee committed:
tuples work

workingjubilee committed:
Single column tables work

workingjubilee committed:
clean up code by deleting it

workingjubilee committed:
use the less-intuitive stable syntax for associated type bounds

workingjubilee committed:
delete some more code

workingjubilee committed:
Delete srf_next

workingjubilee committed:
remove redundant Self: Sized bounds

workingjubilee committed:
Name games

workingjubilee commented on pgrx/src/callconv.rs:
```diff
@@ -134,3 +314,325 @@ fn finish_srf_init<T>(
+        Ret::Once(self)
+    }
+
+    unsafe fn box_return(fcinfo: pg_sys::FunctionCallInfo, ret: Ret<Self>) -> pg_sys::Datum {
+        match ret {
+            Ret::Zero => unsafe { pg_return_null(fcinfo) },
+            Ret::Once(value) => pg_sys::Datum::from(value.to_bits()),
+            Ret::Many(_, _) => unreachable!(),
+        }
+    }
```
This is what I meant about the awkwardness of forcing the scalar type cases into a framework that can support the iterator and composite type cases.

workingjubilee replied:
I resolved this by using a blanket impl.

workingjubilee commented on pgrx/src/callconv.rs:
```diff
@@ -134,3 +314,325 @@ fn finish_srf_init<T>(
+    <T as ReturnShipping>::CallRet: IntoDatum,
+{
+    match ret {
+        Ret::Zero => unsafe { pg_return_null(fcinfo) },
+        Ret::Once(value) => match value.into_datum() {
+            None => unsafe { pg_return_null(fcinfo) },
+            Some(datum) => datum,
+        },
+        Ret::Many(_, _) => unreachable!(),
```
Most of these being should-never-happen cases, effectively.

workingjubilee commented on pgrx/src/iter.rs:
```diff
@@ -155,6 +158,55 @@ impl<'a, T> Iterator for TableIterator<'a, T> {
+    }
+
+    unsafe fn box_return(fcinfo: pg_sys::FunctionCallInfo, ret: Ret<Self>) -> pg_sys::Datum {
+        let value = match ret {
+            Ret::Zero => return empty_srf(fcinfo),
+            Ret::Once(value) => value,
+            Ret::Many(iter, value) => {
+                iter.into_context(fcinfo);
+                value
+            }
```
Probably not even correct, here. We don't have enough tests for this edge-case.

workingjubilee commented on pgrx/src/iter.rs:
```diff
@@ -198,10 +250,38 @@ macro_rules! impl_table_iter {
+                Ret::Once(self)
+            }
+
+            unsafe fn box_return(fcinfo: pg_sys::FunctionCallInfo, ret: Ret<Self>) -> pg_sys::Datum {
+                let value = match ret {
+                    Ret::Zero => return empty_srf(fcinfo),
+                    Ret::Once(value) => value,
+                    Ret::Many(iter, value) => {
+                        iter.into_context(fcinfo);
+                        value
+                    }
+                };
+
+                unsafe {
+                    let fcx = deref_fcx(fcinfo);
+                    let heap_tuple = value.into_heap_tuple((*fcx).tuple_desc);
+                    pg_sys::HeapTupleHeaderGetDatum((*heap_tuple).t_data)
+                }
```
This is the most awkward thing that actually makes me think "dealbreaker, maybe?" until I restructure it: the tuple types shouldn't have to have magical knowledge of the TupleDesc that was set up for them by TableIterator.

workingjubilee replied:
I intend to fix this before we land it but I believe otherwise this is ready for review, I don't want to do much more work on this without feedback.

workingjubilee commented on pgrx/src/iter.rs:
```diff
@@ -198,10 +250,38 @@ macro_rules! impl_table_iter {
+             $($C: ReturnShipping,)*
+             Self: IntoHeapTuple,
+        {
+            type CallRet = Self;
```
so many cases of this...

workingjubilee committed:
Fix UI tests that are suddenly nicer for no reason

workingjubilee committed:
Introduce RetPackage for simple cases

workingjubilee committed:
migrate macros to RetPackage

workingjubilee committed:
Migrate HexInt to RetPackage

workingjubilee committed:
Remove unused code

workingjubilee committed:
Comment on some cases

workingjubilee committed:
s/CallRet/Item/

workingjubilee committed:
rename some macros also

workingjubilee commented on pgrx/src/iter.rs:
```diff
@@ -155,6 +158,56 @@ impl<'a, T> Iterator for TableIterator<'a, T> {
+
+unsafe impl<C> ReturnShipping for (C,)
+where
+    C: ReturnShipping, // so we support TableIterator<'a, (Option<T>,)> as well
```
This is... not ideal, as it traps us in a position where we might accidentally try to return something like `TableIterator<'a, (TableIterator<'b, TupleA>, TableIterator<'c, TupleB>)>`.

workingjubilee replied:
This would be preferable, as it closes off that route:
```suggestion
    C: RetPackage,
```

workingjubilee replied:
I should note that I'm pretty sure that this is currently impossible anyways even with this PR because the schema generator chokes on it, I just would rather it didn't require "the schema generator chokes on it" in order to stop us from doing something silly.

workingjubilee committed:
rework tests to drop Option<Iterator> support

workingjubilee committed:
Convert iter-over-tuple-with-result into result-of-iter

workingjubilee committed:
Use more eager error prop

workingjubilee committed:
Drop support for Option<TableIterator<Tuple>>

workingjubilee commented on pgrx-tests/src/tests/srf_tests.rs:
```diff
@@ -280,13 +280,21 @@ mod tests {
     {
         let oids = vec![1213, 1214, 1232, 1233, 1247, 1249, 1255];
-
-        TableIterator::new(oids.into_iter().map(|oid| {
-            (oid, Spi::get_one(&format!("SELECT CAUSE_AN_ERROR FROM pg_class WHERE oid = {oid}")))
-        }))
+        let result = oids
+            .into_iter()
+            .map(|oid| {
+                Ok((
+                    oid,
+                    Spi::get_one(&format!(
+                        "SELECT CAUSE_AN_ERROR FROM pg_class WHERE oid = {oid}"
+                    ))?,
+                ))
+            })
+            .collect::<Result<Vec<_>, _>>();
+        result.map(|v| TableIterator::new(v))
```
So, this _should_ be `TableIterator<'_, Result<Tuple, Error>>` so that no eager evaluation has to happen. I think I can make that work with a bit more effort. But the schema generator chokes on **that**! And that is another reason why I want it out of our way.

I do not think people should be required to return `TableIterator<'a, (Result, Result, Result, Result, Result, Result)>` if they want to fallibly return a row, but they currently _are_. And I do not think this should be blocked on that.

Asking people to propagate errors to the top of each row return should be fine.

eeeebbbbrrrr replied:
> Asking people to propagate errors to the top of each row return should be fine.

I'm not sure this is a nice thing to ask.  When SPI, for example, is involved, and what you're trying to do is return a TableIterator of a bunch of individual values you somehow got from SPI, it's definitely a lot easier to just return the `Result<T>` (or `Result<Option<T>>`) in whatever tuple position.  Especially if there's quite a few values in the tuple.

This isn't a hill I'm willing to die on, but it'd be nice to support, for example, `TableIterator<(Result<S>, T, Result<U>, Result<V>, W, Result<X>, Y, Result<Z>)>`.  Then we'll panic whenever we get to unwrapping that result and the user doesn't have to worry about it.

workingjubilee replied:
@eeeebbbbrrrr Honestly, we'll never have good UX until we expand to supporting actual `struct` arguments, really.

eeeebbbbrrrr replied:
What does "actual `struct` arguments" mean?

workingjubilee replied:
TableIterator should be able to return a struct that describes the row in question, but this requires being able to use traits like these to return fairly arbitrary generic types. It doesn't work as long as we rely on parsing.

workingjubilee committed:
Stop hiding our crimes from rustfmt

workingjubilee commented on pgrx-examples/spi_srf/src/lib.rs:
```diff
@@ -107,19 +109,17 @@ mod tests {
-        results.push((Ok(Some("Lassie".to_string())), 11, Ok(Some("Collie".to_string())), 77));
-        results.push((Ok(Some("Scooby".to_string())), 13, Ok(Some("Great Dane".to_string())), 91));
-        results.push((Ok(Some("Moomba".to_string())), 15, Ok(Some("Labrador".to_string())), 105));
+        let mut results = Vec::new();
```
An advantage of using a somewhat more sensible return type is that you don't have this sort of thing happen...

workingjubilee committed:
tiny cleanups

workingjubilee reviewed (commented):
As the iterator conveniences are basically independent, I'll rebase this on #1703 after that lands.

workingjubilee committed:
Remove is_option and is_result from ReturnEntity

workingjubilee commented:
This PR opens a LOT of pure simplifications like 9dd8f5fd055748bf6cd3891acb85d3f731f5d782, it's just going to take forever to find them all.

workingjubilee commented on pgrx-examples/spi_srf/src/lib.rs:
```diff
@@ -95,9 +95,11 @@ fn filter_by_breed(
-            .collect::<Vec<_>>();
-        Ok(TableIterator::new(filtered))
+            .map(|row| {
+                Ok((row["dog_name"].value()?, row["dog_age"].value()?, row["dog_breed"].value()?))
```
@eeeebbbbrrrr it's certainly not less code but in most cases it just looks like going "hello? I? have? a? question? if that's okay?"

eeeebbbbrrrr replied:
It's not terrible, but I'm sure you agree that returning a Result from a closure is awkward at best.  In this case we `.collect::<Result<Vec<_>>>()` so it's not too terrible.  

I prefer the left side of this diff, but like I said elsewhere, I'm not willing to die on this hill.

workingjubilee replied:
yeah, the left side is a bit nicer, but that type signature... that's what I really have it in for. so yeah, if I didn't believe it would open up better possibilities in short order, I wouldn't do it.

eeeebbbbrrrr replied:
> that's what I really have it in for

Yeah, that's why I'm not really gonna fight on this one.  I hate typing out all that mess too.  But once the return signature has been typed out, it makes it a lot easier to just write some code and return some values.

NotGyro replied:
Many question marks is much-preferable to having excessively-verbose nested type signatures, yeah.

workingjubilee commented on pgrx/src/callconv.rs:
```diff
@@ -9,96 +9,280 @@
+    unsafe fn ret_from_context(fcinfo: pg_sys::FunctionCallInfo) -> Ret<Self> {
+        let fcx = deref_fcx(fcinfo);
+        // SAFETY: fcx.user_fctx was set earlier, immediately before or in a prior call
+        let iter = &mut *(*fcx).user_fctx.cast::<TableIterator<T>>();
```
uh that is extremely not correct
```suggestion
        let iter = &mut *(*fcx).user_fctx.cast::<SetOfIterator<'_, T>>();
```

workingjubilee committed:
Fix wildly unsound type cast

workingjubilee committed:
Move impl ReturnShipping for *Iterator into pgrx::iter

purely administrative change

workingjubilee committed:
Remove clippy warning that no longer applies

workingjubilee commented:
I moved the implementations for the iterator types to live next to the iterators. It seemed sensical for some reason.

workingjubilee committed:
golf some lines away to see if GH will ever autorender callconv.rs

workingjubilee committed:
Rename ReturnShipping to RetAbi, etc.

eeeebbbbrrrr commented:
Ya know. What if we called it…

I’m kidding. This is great. Thanks, @workingjubilee

workingjubilee committed:
Remove psychic powers from tuple returns

They now know how to create their own tupdesc in the context.

workingjubilee commented:
Resolved the "tuples have psychic powers" issue by splitting out their knowledge of how to add a tupdesc into a new fn.

workingjubilee committed:
Save the argument ABI change for a later PR

workingjubilee committed:
Add some docs

workingjubilee committed:
impl SetOfIterator<'_, T> with TableIterator<'_, (T,)>

workingjubilee committed:
Break one last test

workingjubilee commented on pgrx-tests/src/tests/srf_tests.rs:
```diff
@@ -298,12 +298,15 @@ mod tests {
         let oids = vec![1213, 1214, 1232, 1233, 1247, 1249, 1255];
-
-        SetOfIterator::new(oids.into_iter().map(|oid| {
-            Spi::get_one(&format!("SELECT CAUSE_AN_ERROR FROM pg_class WHERE oid = {oid}"))
-        }))
+        let result = oids
+            .into_iter()
+            .map(|oid| {
+                Spi::get_one(&format!("SELECT CAUSE_AN_ERROR FROM pg_class WHERE oid = {oid}"))
+            })
+            .collect::<Result<Vec<Option<_>>, _>>();
+        result.map(SetOfIterator::new)
```
Breaking this one annoys me much more than the TableIterator nesting case but I want to make sure that SetOfIterator has exactly as stringent requirements, currently, as `TableIterator<'a, (T,)>`, because it will make things much worse down the line if they meaningfully differ.

workingjubilee committed:
remove the fixed fixme

NotGyro reviewed (approved):
Looks good to me. This really does make code working with the returns from Rust to Postgres much cleaner and more elegant. 

I had some confusion about the SetOfIterator type, but it was cleared up in a company meeting. The single-item case behaves differently than the iterator case for legacy Postgres reasons, which is why SetOfIterator can be one item and/or an iterator.

workingjubilee committed:
Explain history and ABI of 1-column RETURNS TABLE

workingjubilee committed:
Rename TableIterator generic to Row?

NotGyro reviewed (approved):
Thanks for the documentation! This explains it neatly.

workingjubilee committed:
Make Ret into an associated type

This allows hiding the details of the type so it cannot be constructed
nor meddled with externally, and simplifies many actual cases.

workingjubilee commented:
This last commit makes the "Ret" type into an associated type and thus makes the iterators eat all the complexity of their return ABI as now only they have to know about how to drive themselves forward. `Result` does not have to think about them, and the implementation details are once more private to iterators and fully hidden behind the trait. This prevents anyone from trying to make any of the **many** bug-inducing nonsensical calls that were technically allowed at first, and probably made the code have to carry around way more panic branches.

workingjubilee committed:
Consistently order the Step variants

idk it just bugged me.

workingjubilee committed:
Add some more docs

workingjubilee committed:
more docs more

workingjubilee committed:
lint the new code with deny(unsafe_op_in_unsafe_fn)

workingjubilee commented:
The last few commits were just cleanup.
